client/components/review.js

```diff
@@ -11,9 +11,106 @@ import { Markdown } from '@ukhomeoffice/asl-components';
+ */
+hasDatabaseChange() {
+    const normaliseValue = value => {
+        if (value === null || value === undefined || (Array.isArray(value) && value.length === 0)) {
+            // Normalise `null`, `undefined`, and empty arrays to a consistent empty state
+            return null;
+        }
+        if (Array.isArray(value)) {
+            // Sort arrays to ensure order doesn't affect comparison
+            return value.sort();
+        }
+        if (typeof value === 'object' && value !== null) {
+            // Ensure consistent ordering of object keys for generic objects
+            return JSON.stringify(value, Object.keys(value).sort());
+        }
+        return value; // Return all other values as-is
+    };
+
+    const normaliseDuration = value => {
+    if (value && typeof value === 'object' && ('years' in value || 'months' in value)) {
+        // Normalise duration objects specifically
+        return {
+            years: value.years ?? 5, // Default to 5 if undefined or null
+            months: value.months ?? 0 // Default to 0 if undefined or null
+        };
+    }
+
+      // Default for completely missing duration
+      return { years: 5, months: 0 };
+  };
+
+
+    const { name, storedValue, currentValue, values } = this.props;
+
+    // Adjust storedValue for duration fields
+    const adjustedStoredValue =
+        name === 'duration' ? normaliseDuration(storedValue) : storedValue;
+
+    // Determine the actual current value for specific fields
+    const actualCurrentValue = currentValue !== undefined && currentValue !== null
+        ? currentValue
+        : values[name] !== undefined
+        ? values[name]
+        : null; // Fallback to `null` if no value is provided
+
+    // Normalise stored and current values
+    const normalisedStoredValue =
+        name === 'duration' ? normaliseDuration(adjustedStoredValue) : normaliseValue(adjustedStoredValue);
+    const normalisedCurrentValue =
+        name === 'duration' ? normaliseDuration(actualCurrentValue) : normaliseValue(actualCurrentValue);
+
+    // Detect changes by comparing normalized values
+    let hasChange = !isEqual(normalisedStoredValue, normalisedCurrentValue);
+
+
+    // Add logic for species-based dynamic fields
+    const species = values?.species || [];
+    if (species.length === 0) {
+      
+    } else {
+        species.forEach(speciesName => {
+            const fieldName = `reduction-quantities-${speciesName}`;
+            const speciesStoredValue = values.storedValue?.[fieldName] || null;
+            const speciesCurrentValue = values[fieldName] || null;
+
+            if (!isEqual(normaliseValue(speciesStoredValue), normaliseValue(speciesCurrentValue))) {
+
+              // any change in dynamic fields sets hasChange to true
+                hasChange = true; 
+            } else {
+                
+            }
+        });
+    }
+
+    return hasChange;
```

Comment:
```suggestion
    if (!isEqual(normalizedStoredValue, normalizedCurrentValue)) {
      return true
    }

    const hasSpeciesChange = (speciesName) => {
      const fieldName = `reduction-quantities-${speciesName}`;
      const speciesStoredValue = values.storedValue?.[fieldName] || null;
      const speciesCurrentValue = values[fieldName] || null;

      return !isEqual(normalizeValue(speciesStoredValue), normalizeValue(speciesCurrentValue));
    }

    const species = values?.species || [];
    return species.find(species => hasSpeciesChange(species)) != null;
```

Also future work: create a ticket to move field specific overrides like this to the schema for that field.

Reply:
Code refactored